File: crates/nerve-tmux-surface/src/status.rs
//! Display derivation, mirrored from Swift. Six painted hues: waiting shares
//! attention; monitor is purple (watching a stream, not executing).
//!
//! The hub deliberately publishes no derived `status`
//! (`crates/nerve-hub/src/model/job.rs:4`) — it stores facets, surfaces paint
//! them — so this transcription of `Nerve/Nerve/Models/Subject.swift:38-100` is
//! the only path to a class, and `tests/status.rs` pins it line by line.
//!
//! Derivation reads structured facets only. No free text ever classifies
//! (CLAUDE.md invariant 3): a `current.summary` of "build failed!!" on a
//! healthy job is a running job with a dramatic summary.

use crate::frame::{AttentionLevel, Health, JobView, Lifecycle, Outcome};

/// `Subject.swift:46-47` — the job is blocked on something else.
///
/// `failure` is a *wait* reason: the job is blocked on a dependency that
/// failed. `outcome` is what says the job itself failed.
const WAIT_REASONS: [&str; 9] = [
    "resource",
    "dependency",
    "queue",
    "system",
    "lock",
    "throttle",
    "rate",
    "capacity",
    "failure",
];

/// `Subject.swift:58` — the job is blocked on a human.
const ASK_REASONS: [&str; 6] = [
    "input",
    "approval",
    "auth",
    "permission",
    "decision",
    "elicitation",
];

/// `Subject.swift:82` — the job is doing work of its own.
const BUSY_KINDS: [&str; 4] = ["subagent", "tool", "thinking", "info"];

/// `Subject.swift:92` — open, but no turn of its own under way.
const IDLE_KINDS: [&str; 3] = ["idle", "starting", "booting"];

/// Derived display status.
///
/// Variants are declared in priority order, so the derived `Ord` *is* the
/// priority (`CoreTypes.swift:244-267`):
/// problem > attention > waiting > running > monitor > success > inactive.
///
/// `Waiting` stays in the enum for that order, the `{waiting}` summary token,
/// and saved palettes. Derivation never returns it: those jobs are Attention
/// so the user learns one “needs a look” color, not two.
#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum StatusClass {
    Problem,
    Attention,
    Waiting,
    Running,
    Monitor,
    Success,
    Inactive,
}

impl StatusClass {
    /// Every class, most urgent first.
    pub const ALL: [StatusClass; 7] = [
        Self::Problem,
        Self::Attention,
        Self::Waiting,
        Self::Running,
        Self::Monitor,
        Self::Success,
        Self::Inactive,
    ];

    /// Lower-case spelling, equal to Swift's `Status.rawValue`.
    pub fn label(self) -> &'static str {
        match self {
            Self::Problem => "problem",
            Self::Attention => "attention",
            Self::Waiting => "waiting",
            Self::Running => "running",
            Self::Monitor => "monitor",
            Self::Success => "success",
            Self::Inactive => "inactive",
        }
    }

    /// The one derivation, ported line for line from `Subject.swift:38-100`.
    ///
    /// The `:NN` comments are that file's line numbers. The three helpers below
    /// are its three `switch`/branch groups — the same grouping `tests/status.rs`
    /// pins section by section — and each answers `None` for exactly the cases
    /// Swift falls through on.
    pub fn of(job: &JobView) -> Self {
        // :39 / :40
        if job.outcome == Some(Outcome::Failure) {
            return Self::Problem;
        }
        if job.health == Health::Unresponsive {
            return Self::Problem;
        }

        if let Some(class) = Self::from_attention(job) {
            return class;
        }
        if let Some(class) = Self::from_stage(job) {
            return class;
        }
        if let Some(class) = Self::from_activity(job) {
            return class;
        }

        // :98 / :99
        if job.lifecycle == Lifecycle::Active {
            Self::Running
        } else {
            Self::Inactive
        }
    }

    /// `Subject.swift:43-66` — what the job is asking of a human.
    fn from_attention(job: &JobView) -> Option<Self> {
        // :43 — elevated attention is attention, never running.
        if job.attention.level >= AttentionLevel::Suggested {
            return Some(Self::Attention);
        }

        // Informational needs a reason to say anything at all.
        if job.attention.level >= AttentionLevel::Informational {
            let reason = Self::reason(job)?;
            if names(reason, &ASK_REASONS) || names(reason, &WAIT_REASONS) {
                return Some(Self::Attention);
            }
        }
        // :64 — anything else falls through to the stage ladder.
        None
    }

    /// `Subject.swift:68-79` — the stage the job as a whole is in.
    fn from_stage(job: &JobView) -> Option<Self> {
        // :68
        if job.lifecycle == Lifecycle::Ended {
            return Some(
                if matches!(job.outcome, Some(Outcome::Success | Outcome::Partial)) {
                    Self::Success // :69
                } else {
                    Self::Inactive // :70
                },
            );
        }
        // :72 / :73 / :74
        if matches!(job.lifecycle, Lifecycle::Suspended | Lifecycle::Unknown) {
            return Some(Self::Inactive);
        }
        if matches!(job.lifecycle, Lifecycle::Pending | Lifecycle::Created) {
            return Some(Self::Attention);
        }
        if job.health == Health::Degraded {
            return Some(Self::Attention);
        }

        // :77 — open session, partial outcome: a monitor holding the stream.
        if job.lifecycle == Lifecycle::Active && job.outcome == Some(Outcome::Partial) {
            return Some(Self::Monitor);
        }
        None
    }

    /// `Subject.swift:81-96` — what the job is doing right now.
    ///
    /// `current.type` is producer vocabulary, not an enum: a spelling this
    /// table never heard of decides nothing (`:94`).
    fn from_activity(job: &JobView) -> Option<Self> {
        let kind = job.current.as_ref()?.kind.as_str();
        // :82 — shell / subagent still running is Running, never Attention.
        if names(kind, &BUSY_KINDS) {
            // :84
            return (job.lifecycle == Lifecycle::Active).then_some(Self::Running);
        }
        // :86 — watching a background stream, not executing.
        if names(kind, &["monitor"]) {
            return Some(Self::Monitor);
        }
        if names(kind, &["waiting"]) {
            return Some(Self::Attention);
        }
        // :92 — `starting` is Ready (open, no turn yet); never Running.
        if names(kind, &IDLE_KINDS) {
            return Some(Self::Inactive);
        }
        None
    }

    /// `Subject.swift:44` / `:56` — the reason, as the producer spelled it.
    ///
    /// Not lower-cased here: matching is case-insensitive
    /// ([`names`]), and this runs once per job per repaint — the sidebar
    /// derives a class for the filter bar, the section list and every row.
    fn reason(job: &JobView) -> Option<&str> {
        job.attention.reason.as_deref()
    }
}

/// Whether producer vocabulary `value` is one of `set`, ignoring case.
///
/// The tables are ASCII, so this is the whole of "lower-case it and compare"
/// without the allocation that phrasing implies.
fn names(value: &str, set: &[&str]) -> bool {
    let value = value.trim();
    set.iter().any(|known| known.eq_ignore_ascii_case(value))
}

File: vsc-ext/src/model/status.ts
/**
 * Display derivation, mirrored from Swift (`Subject.swift:38-100`) and the
 * tmux transcription (`crates/nerve-tmux-surface/src/status.rs`).
 *
 * The hub publishes no derived `status`. Surfaces paint facets. Free text
 * never classifies (CLAUDE.md invariant 3).
 */

import type { AttentionLevel, Job, Status } from "./job";
export { STATUS_PRIORITY } from "./job";

const WAIT_REASONS = [
  "resource",
  "dependency",
  "queue",
  "system",
  "lock",
  "throttle",
  "rate",
  "capacity",
  "failure",
] as const;

/** Ask reasons — human should return to the agent UI (interruptible channel). */
export const ASK_REASONS = [
  "input",
  "approval",
  "auth",
  "permission",
  "decision",
  "elicitation",
  "review",
] as const;

const BUSY_KINDS = ["subagent", "tool", "thinking", "info"] as const;
const IDLE_KINDS = ["idle", "starting", "booting"] as const;

const ATTENTION_RANK: Record<AttentionLevel, number> = {
  none: 0,
  informational: 1,
  suggested: 2,
  required: 3,
  urgent: 4,
};

function names(value: string | undefined, set: readonly string[]): boolean {
  if (!value) return false;
  const trimmed = value.trim();
  return set.some((known) => known.toLowerCase() === trimmed.toLowerCase());
}

function fromAttention(job: Job): Status | undefined {
  if (job.lifecycle === "active" && names(job.current?.type, BUSY_KINDS)) return undefined;
  if (ATTENTION_RANK[job.attention.level] >= ATTENTION_RANK.informational) {
    if (names(job.attention.reason, WAIT_REASONS)) return "waiting";
    if (names(job.attention.reason, ASK_REASONS) || ATTENTION_RANK[job.attention.level] >= ATTENTION_RANK.suggested) return "attention";
  }
  return undefined;
}

function fromStage(job: Job): Status | undefined {
  if (job.lifecycle === "ended") {
    if (job.outcome === "success" || job.outcome === "partial") return "success";
    return "inactive";
  }
  if (job.lifecycle === "suspended" || job.lifecycle === "unknown") {
    return "inactive";
  }
  if (job.lifecycle === "pending" || job.lifecycle === "created") {
    return "waiting";
  }
  if (job.health === "degraded") return "problem";
  if (job.lifecycle === "active" && job.outcome === "partial") return "monitor";
  return undefined;
}

function fromActivity(job: Job): Status | undefined {
  const kind = job.current?.type;
  if (!kind) return undefined;
  if (names(kind, BUSY_KINDS)) {
    return job.lifecycle === "active" ? "running" : undefined;
  }
  if (names(kind, ["monitor"])) return "monitor";
  if (names(kind, ["completed"])) return "success";
  if (names(kind, ["waiting"])) return "waiting";
  if (names(kind, IDLE_KINDS)) return "inactive";
  return undefined;
}

export function statusOf(job: Job): Status {
  if (job.outcome === "failure") return "problem";
  if (job.health === "unresponsive") return "problem";

  if (job.lifecycle === "ended") return fromStage(job)!;
  const attention = fromAttention(job);
  if (attention) return attention;
  const stage = fromStage(job);
  if (stage) return stage;
  const activity = fromActivity(job);
  if (activity) return activity;

  return job.lifecycle === "active" ? "running" : "inactive";
}

export function statusTitle(status: Status): string {
  switch (status) {
    case "problem":
      return "Problem";
    case "attention":
      return "Attention";
    case "waiting":
      return "Waiting";
    case "running":
      return "Running";
    case "monitor":
      return "Monitor";
    case "success":
      return "Success";
    case "inactive":
      return "Inactive";
  }
}

export function statusRank(status: Status): number {
  return ["problem", "attention", "waiting", "running", "monitor", "success", "inactive"].indexOf(
    status,
  );
}

/** Worst (most urgent) painted status in a set. Empty → inactive. */
export function worstStatus(statuses: readonly Status[]): Status {
  if (statuses.length === 0) return "inactive";
  return statuses.reduce((worst, next) =>
    statusRank(next) < statusRank(worst) ? next : worst,
  );
}

export function isAskReason(reason: string | undefined): boolean {
  return names(reason, ASK_REASONS);
}

/** Ask + elevated enough to interrupt (`level ≥ suggested`). */
export function isAskElevated(job: Job): boolean {
  return (
    isAskReason(job.attention.reason) &&
    ATTENTION_RANK[job.attention.level] >= ATTENTION_RANK.suggested
  );
}

/** Whether an Ask interrupt should fire for this transition (upgrade or first sight). */
export function shouldNotifyAsk(previous: Job | undefined, next: Job): boolean {
  if (!isAskElevated(next)) return false;
  if (!previous) return true;
  if (!isAskElevated(previous)) return true;
  return ATTENTION_RANK[next.attention.level] > ATTENTION_RANK[previous.attention.level];
}

export interface AskCopy {
  title: string;
  body: string;
}

/** Gentle toast/banner copy. Prefers producer title/summary; never panics. */
export function askCopy(job: Job, forceUrgentTone = false): AskCopy {
  const reason = (job.attention.reason ?? "input").trim().toLowerCase();
  let fallbackTitle: string;
  switch (reason) {
    case "approval":
    case "permission":
    case "auth":
      fallbackTitle = `Approval needed: ${job.name}`;
      break;
    case "review":
      fallbackTitle = `A review is waiting: ${job.name}`;
      break;
    default:
      fallbackTitle = `Your turn: ${job.name}`;
  }
  let fallbackBody: string;
  if (forceUrgentTone || job.attention.level === "urgent") {
    fallbackBody = "Please return when you can — continue in the agent";
  } else {
    switch (reason) {
      case "approval":
      case "permission":
      case "auth":
        fallbackBody = "Return to the agent to approve";
        break;
      case "review":
        fallbackBody = "Return to the agent when you are ready";
        break;
      default:
        fallbackBody = "Ready when you are — return to the agent to continue";
    }
  }
  const title = job.attention.title?.trim();
  const summary = job.attention.summary?.trim();
  const current = job.current?.summary?.trim();
  return {
    title: title || fallbackTitle,
    body: summary || current || fallbackBody,
  };
}
